server/src/main/java/io/crate/types/TypeCompatibility.java

```diff
@@ -94,8 +94,20 @@
         if (fromTypeParameters.size() != toTypeParameters.size()) {
             if (fromType.id() == ObjectType.ID && toType.id() == ObjectType.ID) {
                 return fromTypeParameters.size() > toTypeParameters.size() ? fromType : toType;
+            } else if (fromType.id() == StringType.ID && toType.id() == StringType.ID) {
+                if (((StringType) fromType).unbound() || ((StringType) toType).unbound()) {
+                    return StringType.INSTANCE;
+                }
             }
             return null;
+        } else {
+            if (fromType.id() == StringType.ID && toType.id() == StringType.ID) {
+                var highestLengthLimit = Math.max(
+                    ((StringType) fromType).lengthLimit(),
+                    ((StringType) toType).lengthLimit()
+                );
+                return StringType.of(highestLengthLimit);
```

Comment:
It would also be possible to return the instance with the higher length to avoid creating a new one.
I doubt it has any significant impact, but the logic wouldn't really be more complicated either.